### PR TITLE
Exporters refactor.

### PR DESCRIPTION
**Pull request by MrK191**

### CODE REVIEW

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java [hunk 1]
```diff
@@ -1,388 +1,139 @@
...
-                clearCachedExport(dataset, formatName);
-            }
+    /**
+     * Exports datasetVersion with given exporter.
+     *
+     * @return {@code Error} if exporting failed or exporter was not found in the list of exporters.
+     * <p>
+     * {@code InputStream} if exporting was an success.
+     */
+    public Either<DataverseError, InputStream> exportDatasetVersion(DatasetVersion datasetVersion, ExporterConstant exporter) {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java]: IMO we should get rid of this version of method. ExportService is not supporting real streaming anyway. Everything is already computed at the time when method returns value.

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Datasets.java [hunk 2]
```diff
@@ -227,25 +234,29 @@ public Response getDataset(@PathParam("id") String id) {
...
 
-            InputStream is = instance.getExport(dataset, exporter);
+        Dataset dataset = datasetService.findByGlobalId(persistentId);
+        if (dataset == null) {
+            return error(Response.Status.NOT_FOUND, "A dataset with the persistentId " + persistentId + " could not be found.");
+        }
 
-            String mediaType = instance.getMediaType(exporter);
+        Either<DataverseError, InputStream> exportedDataset = exportService.exportDatasetVersion(dataset.getReleasedVersion(),
+                                                                                                 exporterConstant.get());
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/api/Datasets.java]: Getting a string version of export should be enough here

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/Xrecord.java [hunk 3]
```diff
@@ -82,17 +86,15 @@ public void writeToStream(OutputStream outputStream, boolean excludeEmailFromExp
...
                 if (dataset != null && formatName != null) {
-                    InputStream inputStream = null;
-                    try {
-                        inputStream = ExportService.getInstance().getExport(dataset, formatName);
-                    } catch (ExportException ex) {
-                        inputStream = null;
-                    }
+                    Either<DataverseError, InputStream> exportedDataset =
+                            exportService.exportDatasetVersion(dataset.getReleasedVersion(),
+                                                               ExporterConstant.valueOf(formatName));
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/Xrecord.java]: Use of string version of method should be enough. Later you can use outputStream.write(...) or something from IOUtils.write

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/qualifiers/ProductionBean.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Indicates if bean that you wanna inject is bean that is used in production environment.
+ * <p>
+ * Should be only used if there is a problem with ambiguous dependency,
+ * which means that someone extended production bean in test environment.
+ */
+@Qualifier
+@Retention(RUNTIME)
+@Target({TYPE, METHOD, PARAMETER, FIELD})
+public @interface ProductionBean {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/qualifiers/ProductionBean.java:22]: I see that annotation as kind of workaround rather then as final solution.
Now the production code must to know how it is tested (that is main code is dependent on test code). And it should be rather the other way around.
But for now we can leave this.

File: dataverse-webapp/src/main/webapp/editMetadata.xhtml [hunk 5]
```diff
@@ -2,12 +2,95 @@
...
                 xmlns:f="http://java.sun.com/jsf/core"
                 xmlns:ui="http://java.sun.com/jsf/facelets"
-                xmlns:c="http://java.sun.com/jsp/jstl/core"
                 xmlns:p="http://primefaces.org/ui"
                 xmlns:o="http://omnifaces.org/ui"
                 xmlns:jsf="http://xmlns.jcp.org/jsf">
 
     <!--@elvariable id="metadataBlocks" type="java.util.List<edu.harvard.iq.dataverse.MetadataBlock>"-->
+    <!-- View Mode -->
+    <div class="panel-group" jsf:rendered="${empty editMode  or managePage}">
```
Comment (madryk) [inline dataverse-webapp/src/main/webapp/editMetadata.xhtml]: Actually this whole block is rendered by viewMetadata.xhtml now

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/ExporterConstant.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package edu.harvard.iq.dataverse.export;
+
+import java.util.Optional;
+
+public enum ExporterConstant {
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/ExporterConstant.java]: `Constant` part of enum name is to vague imo. What do you think about naming it `ExportType`

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/OpenAireExporter.java [hunk 7]
```diff
@@ -1,26 +1,32 @@
...
+    public OpenAireExporter(boolean excludeEmailFromExport) {
+        this.excludeEmailFromExport = excludeEmailFromExport;
     }
 
+    // -------------------- LOGIC --------------------
+
     @Override
     public String getProviderName() {
-        return "oai_datacite";
+        return "openAire";
```
Comment (madryk) [inline dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/export/OpenAireExporter.java]: It seems like this string has to match the value in ExporterConstant. So it seems smart move to use something like `ExporterConstant.OPENAIRE.toString()` or `ExporterConstant.OPENAIRE.getName()` that will return this string

File: dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/export/ExportServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+        when(datasetFieldService.findControlledVocabularyValueByDatasetFieldTypeAndStrValue(subjectFieldType, "Arts and Humanities", false))
+                .thenReturn(subjectFieldType.getControlledVocabularyValue("arts_and_humanities"));
+
+        when(datasetFieldService.findByNameOpt(eq("depositor"))).thenReturn(depositorFieldType);
+        when(datasetFieldService.findByNameOpt(eq("dateOfDeposit"))).thenReturn(dateOfDepositFieldType);
+    }
+
+    private String readFileToString(String resourcePath) throws IOException, URISyntaxException {
+        byte[] bytes = Files.readAllBytes(Paths.get(getClass().getClassLoader().getResource(resourcePath).toURI()));
+        return new String(bytes);
```
Comment (madryk) [inline dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/export/ExportServiceTest.java]: You can achieve the same effect with `IOUtils.resourceToString(classpath, charset)`

**Review by madryk — CHANGES_REQUESTED** [review]

File: dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/qualifiers/ProductionBean.java (hunk 4 above)
Comment (MrK191) [reply dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/qualifiers/ProductionBean.java:22]: Yeah that's true, in the next task we probably should get rid of test classes that are extending 'production beans' and just mock the behavior normally.

**Review by madryk — APPROVED** [review]

**Pull request merged**